### PR TITLE
Pin to taskbar on Windows

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/24054

* Pin to taskbar from first run or default browser dialog when user want to do it.
* Pin to taskbar from settings.

First run dialog:
2022-09-22_12h58_10
Default browser prompt:
2022-09-22_12h56_47
image
Settings when not pinned:
2022-09-22_23h09_34
Settings already pinned:
image

Resolves 

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1.  Launch Brave with clean profile
2. Click `Set Default` button after checking `pin to taskbar`
3. Check shortcut is pinned to taskbar
4. Load brave://settings/getStarted and check settings shows as `pinned`
5. Unpin from taskbar and reload brave://settings/getStarted and click `Pin` button
6. Check pinned to taskbar
7. Unpin from taskbar and set another non-brave browser as a default browser
8. Launch Brave with clean profile
9. Repeat launch till default browser dialog is launched
10. Click `Set Default` button after checking `pin to taskbar`
11. Check shortcut is pinned

### CODE REVIEW

**Comment by simonhong** [comment]
@rmcfadden3 Can you review above settings' words on Windows?
When not pinned, `Pin to taskbar` label and `Pin` button text.
When pinned, `Brave is already pinned`.  cc @rebron

**Comment by rmcfadden3** [comment]
hi @simonhong — just conferred with @rebron . Your copy looks fine to me. Ship it!

File: browser/ui/views/brave_default_browser_dialog_view.cc [hunk 1]
```diff
@@ -44,27 +47,71 @@ namespace {
...
+  void SetFontList() {
+    gfx::FontList font_list;
+    constexpr int kFontSize = 13;
+    font_list.Derive(kFontSize - font_list.GetFontSize(),
+                     font_list.GetFontStyle(), gfx::Font::Weight::NORMAL);
+    label()->SetFontList(font_list);
+  }
+
+  // views::LabelButton overrides:
+  void AddedToWidget() override {
```
Comment (sangwoo108) [inline browser/ui/views/brave_default_browser_dialog_view.cc:73]: nit: I think OnThemeChanged() is proper for this work in a semantic point of view.

**Review by sangwoo108 — COMMENTED** [review]
Sorry for delayed response.

**Comment by sangwoo108** [comment]
@fallaciousreasoning Do you mind taking a look at this? I think I'm not the best one to review polymer here.

File: browser/resources/settings/getting_started_page/getting_started.js [hunk 2]
```diff
@@ -10,6 +10,10 @@ import '../settings_page/settings_animated_pages.js'
 import '../default_browser_page/default_browser_page.js'
 import '../on_startup_page/on_startup_page.js'
 
+// <if expr="is_win">
+import '../pin_shortcut_page/pin_shortcut_page.js'
+// </if>
```
[the remark this answers is not in the document]
Comment (fallaciousreasoning) [reply browser/resources/settings/getting_started_page/getting_started.js:15]: I think it's basically a conditional include :)

File: browser/resources/settings/pin_shortcut_page/pin_shortcut_page.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
```
Comment (fallaciousreasoning) [inline browser/resources/settings/pin_shortcut_page/pin_shortcut_page.js:1]:
```suggestion
/* Copyright (c) 2022 The Brave Authors. All rights reserved.
```

**Review by fallaciousreasoning — APPROVED** [review]
Polymer stuff LGTM

File: browser/resources/settings/getting_started_page/getting_started.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/resources/settings/getting_started_page/getting_started.js:15]: Yeah, it's not the comment.

**Review by simonhong — COMMENTED** [review]
@sangwoo108 Thanks for review. and I realized that upstream's taskbar api is available now. 
Updated to use it and this feature can be used on win7/8 also.

**Comment by simonhong** [comment]
@sangwoo108 Can you take another look?

File: browser/ui/views/brave_default_browser_dialog_view.cc [hunk 4]
```diff
@@ -44,27 +47,72 @@ namespace {
...
-  NoSnappedBubbleFrameView& operator=(const NoSnappedBubbleFrameView&) = delete;
+  explicit DontAskAgainButton(PressedCallback callback)
+      : LabelButton(std::move(callback)) {
+    SetFontList();
+    SetText(brave_l10n::GetLocalizedResourceUTF16String(
+        IDS_BRAVE_DEFAULT_BROWSER_DIALOG_DONT_ASK));
+  }
+  DontAskAgainButton(const DontAskAgainButton&) = delete;
+  DontAskAgainButton& operator=(const DontAskAgainButton&) = delete;
+
```
Comment (sangwoo108) [inline browser/ui/views/brave_default_browser_dialog_view.cc:62]: nit: ~DontAskAgainButton() override = default;

File: browser/ui/views/brave_first_run_dialog.cc [hunk 5]
```diff
@@ -42,8 +40,38 @@ void ShowBraveFirstRunDialogViews(Profile* profile) {
...
+    constexpr int kFontSize = 14;
+    font_list.Derive(kFontSize - font_list.GetFontSize(),
+                     font_list.GetFontStyle(), gfx::Font::Weight::NORMAL);
+    label()->SetFontList(font_list);
+  }
+};
+
+BEGIN_METADATA(PinShortcutCheckbox, views::Checkbox)
+END_METADATA
+#endif
```
Comment (sangwoo108) [inline browser/ui/views/brave_first_run_dialog.cc:69]:
```suggestion
#endif  // BUILDFLAG(IS_WIN)
```

File: browser/ui/webui/settings/pin_shortcut_handler.cc [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+void PinShortcutHandler::OnCheckShortcutPinState(bool pinned) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+
+  NotifyShortcutPinStateChangeToPage(pinned);
+}
+#endif
+
+void PinShortcutHandler::NotifyShortcutPinStateChangeToPage(bool pinned) {
+  if (IsJavascriptAllowed()) {
```
[the remark this answers is not in the document]
Comment (sangwoo108) [reply browser/ui/webui/settings/pin_shortcut_handler.cc:69]: > // This should never be called from a C++ callback used as a reply for a

Oh, didn't know that. Thanks!

**Review by sangwoo108 — APPROVED** [review]
LGTM with a few nits! 👍

File: browser/ui/views/brave_first_run_dialog.cc (hunk 5 above)
Comment (simonhong) [reply browser/ui/views/brave_first_run_dialog.cc]: Done.

File: browser/ui/views/brave_default_browser_dialog_view.cc (hunk 4 above)
Comment (simonhong) [reply browser/ui/views/brave_default_browser_dialog_view.cc:63]: Added.

**Comment by MadhaviSeelam** [comment]
@sangwoo108 I think `step 9` in the `Test Plan` needs to be modified as `Infobar` replaced `default browser dialog` - https://github.com/brave/brave-browser/issues/25929

cc: @rebron

**Comment by simonhong** [comment]
@MadhaviSeelam Updated test plan in PR and issue. Deleted default browser dialog part.

**Pull request merged**